autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+type AutoretrieveEngine struct {
+	*options
+	lsys ipld.LinkSystem
+
+	entriesChunker *chunker.CachedEntriesChunker
+
+	publisher legs.Publisher
+
+	mhLister provider.MultihashLister
+	cblk     sync.Mutex
+}
+
+// var _ provider.Interface = (*AutoretrieveEngine)(nil)
+
+// New creates a new index provider Engine as the default implementation of provider.Interface.
+// It provides the ability to advertise the availability of a list of multihashes associated to
+// a context ID as a chain of linked advertisements as defined by the indexer node protocol implemented by "storetheindex".
+// Engine internally uses "go-legs", a protocol for propagating and synchronizing changes an IPLD DAG, to publish advertisements.
+// See:
+//  - https://github.com/filecoin-project/storetheindex
+//  - https://github.com/filecoin-project/go-legs
+//
+// Published advertisements are signed using the given private key.
+// The retAddrs corresponds to the endpoints at which the data block associated to the advertised
+// multihashes can be retrieved.
+// Note that if no retAddrs is specified the listen addresses of the given libp2p host are used.
+//
+// The engine also provides the ability to generate advertisements via Engine.NotifyPut and
+// Engine.NotifyRemove as long as a provider.MultihashLister is registered.
+// See: provider.MultihashLister, Engine.RegisterMultihashLister.
+//
+// The engine must be started via Engine.Start before use and discarded via Engine.Shutdown when no longer needed.
+// See: Engine.Start, Engine.Shutdown.
+func New(o ...Option) (*AutoretrieveEngine, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+
+	e := &AutoretrieveEngine{
+		options: opts,
+	}
+
+	e.lsys = e.mkLinkSystem()
+
+	return e, nil
+}
+
+// Start starts the engine by instantiating the internal storage and joins the configured gossipsub
+// topic used for publishing advertisements.
+//
+// The context is used to instantiate the internal LRU cache storage.
+//
+// See: AutoretrieveEngine.Shutdown, chunker.NewCachedEntriesChunker, dtsync.NewPublisherFromExisting.
+func (e *AutoretrieveEngine) Start(ctx context.Context) error {
+	// Create datastore entriesChunker
+	entriesCacheDs := dsn.Wrap(e.ds, datastore.NewKey(linksCachePath))
+	cachedChunker, err := chunker.NewCachedEntriesChunker(ctx, entriesCacheDs, e.entChunkSize, e.entCacheCap)
+	if err != nil {
+		return err
+	}
+
+	if e.purgeCache {
+		err := cachedChunker.Clear(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	e.entriesChunker = cachedChunker
+
+	e.publisher, err = e.newPublisher()
+	if err != nil {
+		log.Errorw("Failed to instantiate legs publisher", "err", err, "kind", e.pubKind)
+		return err
+	}
+
+	// Initialize publisher with latest advertisement CID.
+	adCid, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get latest advertisement cid: %w", err)
+	}
+	if adCid != cid.Undef {
+		if err = e.publisher.SetRoot(ctx, adCid); err != nil {
```

Comment:
for `NoPublisher` wont call to `e.publisher.SetRoot` panic since  `e.publisher` will be nil?

Reply:
Same as above, see https://github.com/filecoin-project/index-provider/blob/dfc4fa4ad02a4f0494b67f7f6c15581c434ef89b/engine/engine.go. Might be worth checking.

Reply:
@gmelodie I just looked, I can see they have fixed my concern here https://github.com/filecoin-project/index-provider/commit/563785de47bf47c8ae5ffc407d6fd7294e219a03, so I think you have to check, else it will panic